Metric reporters (Graphite, InfluxDB)

lbendig opened the pull request:
Re-implementation of the Graphite and InfluxDB reporters that were removed in https://github.com/linkedin/gobblin/pull/429 due to the changes in Metric API.

Graphite:
- Supports sending through UDP and TCP 
- Metric and/or event reporter can be toggled
- Additional parameter `metrics.reporting.graphite.events.value.as.key`: 
  - For those backends that don't allow strings as event values (e.g: InfluxDB's Graphite protocol)
    In this case the event value will be added to the end of the key

InfluxDB:
- Metric and/or event reporter can be toggled
- Only TCP as influxdb-java 2.1 doesn't support UDP yet

Both were tested on InfluxDB 0.9.4.1.

Refactoring:
- Restructured configuration keys for the metric reporters and classes of metric/event names
- Timing event related event names are now in the `TimingEvent` class
- Metric reporting configurations are moved from `MetricReportReport` to a new base class `ConfiguredScheduledReporter`.
  To report non-serialized metrics one can extend `ConfiguredScheduledReporter`, for serialized Avro records `MetricReportReport` can be subclassed
- Duplicate class: MetricGroup  -> gobblin.runtime.MetricGroup was removed

ibuenros commented:
Thanks @lbendig . It is a large PR, I'll try to look at it over the weekend.

lbendig commented:
Thanks @ibuenros !

ibuenros commented on gobblin-metrics/src/main/java/gobblin/metrics/event/MultiPartEvent.java:
```diff
@@ -0,0 +1,87 @@
+  private String eventName;
+  private String[] metadataFields;
+  
+  private static final Map<String, MultiPartEvent> lookup = Maps.newHashMap();
```
Can you make this an `ImmutableMap` instead? Helps avoid memory leaks.

lbendig replied:
Fixed.

lbendig commented:
@ibuenros pls just let me know if you need anything regarding the review of this PR.

lbendig commented:
@ibuenros what do you think, would this PR be a good candidate to add it to Gobblin?

ibuenros commented:
Sorry about the delay (I know it's ridiculous). I promise to look at it by end of weekend.

ibuenros commented on gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java:
```diff
@@ -548,6 +557,155 @@ private void buildKafkaMetricReporter(Properties properties) {
+        Utils.getPropAsBoolean(properties, ConfigurationKeys.METRICS_REPORTING_GRAPHITE_EVENTS_ENABLED_KEY,
+            ConfigurationKeys.DEFAULT_METRICS_REPORTING_GRAPHITE_EVENTS_ENABLED);
+    if (!metricsEnabled && !eventsEnabled) {
+      LOGGER.info("Not reporting metrics and events to Graphite");
```
I know we do it this way right now, but I would prefer instead logging the positive case. E.g. if the graphite metrics are enabled, log "Reporting metrics... Graphite"

lbendig replied:
Fixed. I also changed the way of logging (positive case) for the other reporters.

ibuenros commented on gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java:
```diff
@@ -548,6 +557,155 @@ private void buildKafkaMetricReporter(Properties properties) {
+    }
+    
+    try {
+      Preconditions.checkArgument(properties.containsKey(ConfigurationKeys.METRICS_REPORTING_GRAPHITE_HOSTNAME),
```
I think we should fail if something critical is missing. Better than not getting metrics.

lbendig replied:
I followed the pattern of the other reporters: catch exception, log it and return. Should any issue in the metric reporting bring down the job?

ibuenros commented on gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java:
```diff
@@ -548,6 +557,155 @@ private void buildKafkaMetricReporter(Properties properties) {
+    
+    String url = properties.getProperty(ConfigurationKeys.METRICS_REPORTING_INFLUXDB_URL);
+    String username = properties.getProperty(ConfigurationKeys.METRICS_REPORTING_INFLUXDB_USER);
+    String password = properties.getProperty(ConfigurationKeys.METRICS_REPORTING_INFLUXDB_PASSWORD);
```
Not necessary, but would it be good to allow encryption of this password? (There is a password manager module in Gobblin).

lbendig replied:
Good idea, fixed.

ibuenros commented on gobblin-metrics/src/main/java/gobblin/metrics/graphite/GraphiteEventReporter.java:
```diff
@@ -0,0 +1,231 @@
+    GobblinTrackingEvent nextEvent;
+    try {
+      while (null != (nextEvent = queue.poll())) {
+        pushEvent(nextEvent);
```
can `pushEvent` fail on certain event, failing to flush previous events?

lbendig replied:
I think this can be solved by flushing the events in the catch block.

ibuenros commented:
It looks pretty good in general. A few comments on the Graphite reporter, which also apply to the corresponding methods for InfluxDB.

lbendig commented:
@ibuenros  Thank you for the review! I addressed your comments and also made changes in the metric key naming for the Influx and Graphite reporters as I found that it was impacted by some commits since the initial pull request. Also downgraded influxdb-java to 2.1 because the api went through some major changes in 2.2.

ibuenros commented:
Merging. Thanks @lbendig .

lbendig commented:
Thank you @ibuenros for the review/merge!
